all-the-icons.el

```diff
@@ -1165,6 +1165,14 @@ Return the icon file name if found."
 
 (defconst all-the-icons--lib-dir (file-name-directory (locate-library "all-the-icons")))
 
+(defun all-the-icons-font-size (&optional face)
```

Comment:
all-the-icons never really supported using it in the terminal, and in the case of the term, `frame-char-height` always return 1, why do we need this function at all?

Reply:
This was adapted from `window-default-font-height` but altered to pull the second field instead of the third field from the font info. But I can definitely simplify it.

Reply:
Pushed.

Reply:
You can't always assume slot 2 is defined. Whether slot 2 or 3 is define depends on a lot of things you have to take care of all the cases. I still don't understand why you need this method. How you get the height of the font is highly dependent on where you are going to insert the font. If you need to insert an icon onto the tab line, and it's somehow using a face that's shorter than your `window-default-font-height`, just take the size of the tab line font and pass it to the icon.

https://www.gnu.org/software/emacs/manual/html_node/emacs/Fonts.html#Fonts

Reply:
Although, I still think if this function is needed, you need to take from slot 3 otherwise the tail of the Q will be below the icon.

Reply:
> You can't always assume slot 2 is defined. Whether slot 2 or 3 is define depends on a lot of things, you have to take care of all the cases.

I can add a fallback on slot 3 if slot 2 is 0 or undefined. If slot 3 isn't defined, the previous code wouldn't have worked either.

Although I can't find any cases where slot 2 isn't defined even when I open random fonts:

```elisp
(aref (font-info "-*-Arvo-regular-normal-normal-*-*-*-*-*-*-*-*-*") 2)
```

> Also, I still don't understand why you need this method. How you get the height of the font is highly dependent on where you are going to insert the font. If you need to insert an icon onto the tab line, and it's somehow using a face that's shorter than your window-default-font-height, just take the size of the tab line font and pass it to the icon.

I agree, that's why I'm providing a way to for the user to specify the height. I can't 100% know the correct height because I have no idea where the user is going to insert the icon.

However, we can make a good guess. What I'm trying to do here is to make a _better_ guess than the current code:

- The current code guesses the font "height" of the current window. But that included the ascent/descent and, in my experience, increases the height of lines with icons on them.
- The new code uses the font "size" as the guess, which appears to work better.

To be clear, my issue here isn't the tabline, it's regular buffers (notmuch, ibuffer, completions, etc.).

Reply:
To clear up any confusion, `window-default-font-height` (the previously called function) is defined as follows:

```elisp
(defun window-default-font-height (&optional window)
  "Return height in pixels of WINDOW's default face font.
WINDOW must be a live window and defaults to the selected one.

The return value accounts for any remapping of the default face
font on WINDOW's frame."
  (let* ((window (window-normalize-window window t))
	 (frame (window-frame window))
	 (default-font (face-font 'default frame)))
    ;; Client frames can have the 'display' parameter set like for X
    ;; frames, even though they are TTY frames, so make sure we won't
    ;; be duped by that up front with 'framep'.
    (if (and (not (eq (framep frame) t))
             (display-multi-font-p (frame-parameter frame 'display))
	     (not (string-equal (frame-parameter frame 'font) default-font)))
        (aref (font-info default-font frame) 3)
      (frame-char-height frame))))
```

However:

1. All the logic around getting the frame/window/etc. isn't necessary in this case because we never pass an explicit window.
2. That code checks the target font against the default font as an optimization, from what I can tell, so we can just pull the default height from the frame. Unfortunately, that doesn't work here because we want the font "size", not the font height.

Reply:
I've pushed an extra cautious version. I don't think it's necessary and I'm pretty sure I'm misunderstanding your request, but I figured this would get us closer to being on the same page.

Reply:
The "size" in the `all-the-icons-define-icon` macro actually means the height of the font, the size means the size of one of the sides of the icons, which is always transformed into a square. In Emacs font-spec speak, it also means the height. It's just Emacs calls the height - ascent - decent `pixel-size`.